x/stake/handler.go

```diff
@@ -53,212 +63,245 @@ func EndBlocker(ctx sdk.Context, k Keeper) (ValidatorUpdates []abci.Validator) {
-	}
-	if msg.Bond.Denom != k.GetParams(ctx).BondDenom {
-		return ErrBadBondingDenom(k.codespace).Result()
+		return ErrValidatorAlreadyExists(k.Codespace()).Result()
 	}
-	if ctx.IsCheckTx() {
-		return sdk.Result{}
+	if msg.SelfDelegation.Denom != k.GetParams(ctx).BondDenom {
+		return ErrBadDenom(k.Codespace()).Result()
 	}
 
 	validator := NewValidator(msg.ValidatorAddr, msg.PubKey, msg.Description)
-	k.setValidator(ctx, validator)
-	k.setValidatorByPubKeyIndex(ctx, validator)
-	tags := sdk.NewTags(
-		"action", []byte("createValidator"),
-		"validator", msg.ValidatorAddr.Bytes(),
-		"moniker", []byte(msg.Description.Moniker),
-		"identity", []byte(msg.Description.Identity),
-	)
+	k.SetValidator(ctx, validator)
+	k.SetValidatorByPubKeyIndex(ctx, validator)
 
-	// move coins from the msg.Address account to a (self-bond) delegator account
+	// move coins from the msg.Address account to a (self-delegation) delegator account
 	// the validator account and global shares are updated within here
-	delegateTags, err := delegate(ctx, k, msg.ValidatorAddr, msg.Bond, validator)
+	_, delegation, validator, pool, err := k.Delegate(ctx, msg.ValidatorAddr, msg.SelfDelegation, validator)
+	k.SetPool(ctx, pool)
+	k.SetDelegation(ctx, delegation)
+	k.UpdateValidator(ctx, validator)
 	if err != nil {
 		return err.Result()
 	}
-	tags = tags.AppendTags(delegateTags)
+
+	tags := sdk.NewTags(
+		tags.Action, tags.ActionCreateValidator,
+		tags.DstValidator, msg.ValidatorAddr.Bytes(),
+		tags.Moniker, []byte(msg.Description.Moniker),
+		tags.Identity, []byte(msg.Description.Identity),
+	)
 	return sdk.Result{
 		Tags: tags,
 	}
 }
 
-func handleMsgEditValidator(ctx sdk.Context, msg MsgEditValidator, k Keeper) sdk.Result {
+func handleMsgEditValidator(ctx sdk.Context, msg types.MsgEditValidator, k keeper.Keeper) sdk.Result {
 
 	// validator must already be registered
 	validator, found := k.GetValidator(ctx, msg.ValidatorAddr)
 	if !found {
-		return ErrBadValidatorAddr(k.codespace).Result()
-	}
-	if ctx.IsCheckTx() {
-		return sdk.Result{}
+		return ErrNoValidatorFound(k.Codespace()).Result()
 	}
 
-	// XXX move to types
 	// replace all editable fields (clients should autofill existing values)
-	validator.Description.Moniker = msg.Description.Moniker
-	validator.Description.Identity = msg.Description.Identity
-	validator.Description.Website = msg.Description.Website
-	validator.Description.Details = msg.Description.Details
+	description, err := validator.Description.UpdateDescription(msg.Description)
+	if err != nil {
+		return err.Result()
+	}
+	validator.Description = description
 
-	k.updateValidator(ctx, validator)
+	k.UpdateValidator(ctx, validator)
 	tags := sdk.NewTags(
-		"action", []byte("editValidator"),
-		"validator", msg.ValidatorAddr.Bytes(),
-		"moniker", []byte(msg.Description.Moniker),
-		"identity", []byte(msg.Description.Identity),
+		tags.Action, tags.ActionEditValidator,
+		tags.DstValidator, msg.ValidatorAddr.Bytes(),
```

Comment:
Needs to be passed as `[]byte(msg.ValidatorAddr.String())` so we get the hex-encoded address.

Reply:
addressed for validator (src/dst) and delegators everywhere